src/Store/VimStoreConnector.re

```diff
@@ -631,15 +632,21 @@ let start =
         state,
         pasteIntoEditorAction,
       )
-    | Model.Actions.WildmenuNext
-    | Model.Actions.WildmenuPrevious
-    | Model.Actions.WildmenuSelect(_) =>
+    | Model.Actions.MenuFocusNext
+    | Model.Actions.MenuFocusPrevious
+    | Model.Actions.MenuFocus(_) =>
+      // IFFY: Depends on the ordering of "updater"s
```

Comment:
This 'update-ordering' dependency is tricky for sure. I wonder if there is a better way for us to model this? Does Elm have the same problem? 🤔

Reply:
Elm code typically isn't split up this way, for several reasons: 1) Because it uses "subscriptions" to start and stop continuous processes, which of course is why I've tried to replicate that here. And 2) because using the wildcard pattern is discouraged and exhaustiveness checking on variants to driving development is an important part of the Elm workflow.

I think Evan's talk on "The Life of a File" goes through this workflow, and Richard Feldman's tour of `elm-spa-example` goes into more high-level choices. But the workflow can roughly be summed up as first thinking very hard about the model to make sure you get it right, what states should be representable and which should be impossible to represent (Yaron Minsky has a good talk on this), and what state transitions (ie. actions) are needed. Then just run the compiler until it stops complaining. The types and exhaustiveness checking makes sure you handle all the possible actions and initialize all the values correctly. All that's left then is to render the UI and wire up the events.

If a file seems to get too big to handle (a line of thinking that Evan is very much against, mind you) You'll typically split it up by defining a sub-action type and giving it its own constructor in the main action variant. So just like you have `state.quickmenu`, you might have an action `Quickmenu(Quickmenu.action)`.

To avoid the resulting tedious boilerplate when creation actions, like `dispatch(Quickmenu(MenuSelect(42))`, Elm provides `map` functions, like `Html.map` to map `Html` elements from one message type to another. In F#'s elmish, which also uses a dispatch function, [this is very easy to do with just function composition](https://stackoverflow.com/a/53678466/7943564). Unfortunately Reason does not have a built-in function composition operator or first-class variant constructors.

Reply:
Thanks for the resources and details here! I'll check out the talk and Richard Feldman's example.

> If a file seems to get too big to handle (a line of thinking that Evan is very much against, mind you) You'll typically split it up by defining a sub-action type and giving it its own constructor in the main action variant. So just like you have state.quickmenu, you might have an action Quickmenu(Quickmenu.action).

Interesting. I like this - I could see this being more understandable than our uber `Actions.t`.

> Unfortunately Reason does not have a built-in function composition operator or first-class variant constructors.

This is a bummer. I keep hoping the first-class variant constructors will come with the OCaml compiler.

Reply:
> I keep hoping the first-class variant constructors will come with the OCaml compiler.

As do I, but I wouldn't hold my breath. It was a conscious design decision back in the day, and I think the confusingly ambiguous syntax that resulted makes it difficult to implement in both a non-breaking and non-awkward way.